plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java

```diff
@@ -125,6 +128,124 @@ public void respond(
+      int rangeStart = line.getStartOffset();
+      int rangeEnd = rangeStart + line.getLength();
+
+      ArrayList<org.eclipse.che.ide.api.editor.text.Position> allPositions = new ArrayList<>();
+      List<DiagnosticAnnotation> allAnnotations = new ArrayList<>();
+      int bestOffset = Integer.MAX_VALUE;
+      for (Annotation a : annotations.keySet()) {
+        org.eclipse.che.ide.api.editor.text.Position pos = annotations.get(a);
+        if (pos != null && isInside(pos.offset, rangeStart, rangeEnd)) { // inside our range?
+          allAnnotations.add((DiagnosticAnnotation) a);
+          allPositions.add(pos);
+          bestOffset = processAnnotation(a, pos, invocationLocation, bestOffset);
+        }
+      }
+      if (bestOffset == Integer.MAX_VALUE) {
+        return invocationLocation;
+      }
+      for (int i = 0; i < allPositions.size(); i++) {
+        org.eclipse.che.ide.api.editor.text.Position pos = allPositions.get(i);
+        if (isInside(bestOffset, pos.offset, pos.offset + pos.length)) {
+          resultingDiagnostics.add(allAnnotations.get(i).getDiagnostic());
+        }
+      }
+
+      return bestOffset;
+    } else {
+      // iteration with range never returns anything; need to filter ourselves.
+      // https://github.com/eclipse/che/issues/4338
+      resultingDiagnostics.addAll(
+          annotations
+              .entrySet()
+              .stream()
+              .filter((e) -> e.getValue().overlapsWith(range.getStartOffset(), range.getLength()))
+              .map(Entry::getKey)
+              .map(a -> (DiagnosticAnnotation) a)
+              .map(DiagnosticAnnotation::getDiagnostic)
+              .collect(Collectors.toList()));
+    }
+    return invocationLocation;
+  }
+
+  /**
+   * Tells is the offset is inside the (inclusive) range defined by start-end.
+   *
+   * @param offset the offset
+   * @param start the start of the range
+   * @param end the end of the range
+   * @return true iff offset is inside
+   */
+  private boolean isInside(int offset, int start, int end) {
+    return offset == start
+        || offset == end
+        || (offset > start && offset < end); // make sure to handle 0-length ranges
+  }
+
+  private int processAnnotation(
+      Annotation annot,
+      org.eclipse.che.ide.api.editor.text.Position pos,
+      int invocationLocation,
+      int bestOffset) {
+    final int posBegin = pos.offset;
+    final int posEnd = posBegin + pos.length;
+    if (isInside(invocationLocation, posBegin, posEnd)) { // covers invocation location?
+      return invocationLocation;
+    } else if (bestOffset != invocationLocation) {
+      final int newClosestPosition = computeBestOffset(posBegin, invocationLocation, bestOffset);
+      if (newClosestPosition != -1) {
+        if (newClosestPosition != bestOffset) { // new best
+          // Can't use JavaAnnotationUtil.hasCorrections() to see if there are
```

Comment:
However I admit that in some cases there is indeed no better way to explain the idea than to use a proper comment :-D

Reply:
In this case it is my own comment on that I wasn't able to replace JDT realization with something more or less similar in the LSP one. And as such this potentially can be a reason of why LSP realization works differently in some cases (and i believe there will be such cases)